docroot/modules/humsci/hs_config_readonly/src/EventSubscriber/ConfigReadOnlyEventSubscriber.php

```diff
@@ -0,0 +1,222 @@
+ * configurations provided by modules.
+ *
+ * @package Drupal\hs_config_readonly\EventSubscriber
+ */
+class ConfigReadOnlyEventSubscriber implements EventSubscriberInterface {
+
+  use ConfigReadonlyWhitelistTrait;
+
+  /**
+   * Installed configuration service.
+   *
+   * @var \Drupal\Core\Config\ExtensionInstallStorage
+   */
+  protected $extensionConfigStorage;
+
+  /**
+   * Optional configuration service.
+   *
+   * @var \Drupal\Core\Config\ExtensionInstallStorage
+   */
+  protected $extensionOptionalConfigStorage;
+
+  /**
+   * Ignore the configurations form these modules.
+   *
+   * @var array
+   */
+  protected $excludedModules = ['system'];
+
+  /**
+   * Root path of Drupal.
+   *
+   * @var string
+   */
+  protected $drupalRoot;
+
+  /**
+   * Form ids to mark as read only.
+   */
+  protected $readOnlyFormIds = [
+    'config_single_import_form',
+    'system_modules',
+    'system_modules_uninstall',
+    'user_admin_permissions',
+  ];
+
+  /**
+   * {@inheritdoc}
+   */
+  public function __construct($drupal_root, ModuleHandlerInterface $module_handler, ExtensionInstallStorage $extension_config_storage, ExtensionInstallStorage $extension_optional_config_storage) {
+    $this->drupalRoot = $drupal_root;
+    $this->moduleHandler = $module_handler;
+    $this->extensionConfigStorage = $extension_config_storage;
+    $this->extensionOptionalConfigStorage = $extension_optional_config_storage;
+  }
+
+  public function onFormAlter(ReadOnlyFormEvent $event) {
+    // Check if the form is a ConfigFormBase or a ConfigEntityListBuilder.
+    $form_object = $event->getFormState()->getFormObject();
+    $mark_form_read_only = $form_object instanceof ConfigFormBase;
+
+    if (!$mark_form_read_only) {
+      $mark_form_read_only = in_array($form_object->getFormId(), $this->readOnlyFormIds);
+    }
+
+    // Check if the form is an EntityFormInterface and entity is a config
+    // entity.
+    if (!$mark_form_read_only && $form_object instanceof EntityFormInterface) {
+      $entity = $form_object->getEntity();
+      $mark_form_read_only = $entity instanceof ConfigEntityInterface;
+    }
+
+    // Don't block particular patterns.
+    if ($mark_form_read_only && $form_object instanceof EntityFormInterface) {
+      $entity = $form_object->getEntity();
+      $name = $entity->getConfigDependencyName();
+
+      // Block config from a module.
+      $mark_form_read_only = $this->configIsEditable([$name]);
+
+      // Unless its defined as a whiteist pattern.
+      if ($this->matchesWhitelistPattern($name)) {
+        $mark_form_read_only = FALSE;
+      }
+    }
+
+    // Config forms.
+    if ($mark_form_read_only && $form_object instanceof ConfigFormBase) {
+      $editable_config = $this->getEditableConfigNames($form_object);
+      $mark_form_read_only = $this->configIsEditable($editable_config);
+    }
+
+    if ($mark_form_read_only) {
+      $event->markFormReadOnly();
```

Comment:
This is a handy little method.

Reply:
yeah i wish it had a `$event->markFormEditable()` or something. cuz then i wouldn't need to replace the service. I could just create another event subscriber